tensorboard/plugins/audio/tf_audio_dashboard/tf-audio-dashboard.ts

```diff
@@ -0,0 +1,188 @@
+              <a
+                href="https://github.com/tensorflow/tensorboard/blob/master/README.md#my-tensorboard-isnt-showing-any-data-whats-wrong"
+                >the section of the README devoted to missing data problems</a
+              >
+              and consider filing an issue on GitHub.
+            </p>
+          </div>
+        </template>
+        <template is="dom-if" if="[[!_dataNotFound]]">
+          <tf-tag-filterer tag-filter="{{_tagFilter}}"></tf-tag-filterer>
+          <template is="dom-repeat" items="[[_categories]]" as="category">
+            <tf-category-paginated-view
+              category="[[category]]"
+              initial-opened="[[_shouldOpen(index)]]"
+            >
+              <template>
+                <tf-audio-loader
+                  active="[[active]]"
+                  run="[[item.run]]"
+                  tag="[[item.tag]]"
+                  sample="[[item.sample]]"
+                  total-samples="[[item.totalSamples]]"
+                  tag-metadata="[[_tagMetadata(_runToTagInfo, item.run, item.tag)]]"
+                  request-manager="[[_requestManager]]"
+                ></tf-audio-loader>
+              </template>
+            </tf-category-paginated-view>
+          </template>
+        </template>
+      </div>
+    </tf-dashboard-layout>
+    <style include="dashboard-style"></style>
+    <style>
+      .no-data-warning {
+        max-width: 540px;
+        margin: 80px auto 0 auto;
+      }
+    </style>
+  `;
+  @property({type: Array})
+  _selectedRuns: string[] = [];
+  @property({type: Object})
+  _runToTagInfo: AudioBackendTagsResponse | null = null;
+  @property({type: Boolean})
+  _dataNotFound: boolean = true;
+  @property({type: String})
+  _tagFilter: string = '';
+  @property({type: Object})
+  _requestManager = new tf_backend.RequestManager();
+  ready() {
+    super.ready();
+    this.reload();
+  }
+  reload() {
+    this._fetchTags().then(() => {
+      this._reloadAudio();
+    });
+  }
+  _fetchTags() {
+    const url = tf_backend.getRouter().pluginRoute('audio', '/tags');
+    return this._requestManager
+      .request(url)
+      .then((runToTagInfo: AudioBackendTagsResponse) => {
+        if (_.isEqual(runToTagInfo, this._runToTagInfo)) {
+          // No need to update anything if there are no changes.
+          return;
+        }
+        const runToTag = _.mapValues(runToTagInfo, (x) => Object.keys(x));
+        const tags = tf_backend.getTags(runToTag);
+        this.set('_dataNotFound', tags.length === 0);
+        this.set('_runToTagInfo', runToTagInfo);
+      });
+  }
+  _reloadAudio() {
+    this.root!.querySelectorAll('tf-audio-loader').forEach(
+      (audio: TfAudioLoader) => {
+        audio.reload();
+      }
+    );
+  }
+  _shouldOpen(index: number) {
+    return index <= 2;
+  }
+  @computed('_runToTagInfo', '_selectedRuns', '_tagFilter')
+  get _categories() {
```

Comment:
can you do me a favor and specify the return type for this? This method is a bit more involved and I cannot picture what the return type should be.

Reply:
Done.  Yeah, I should have annotated this when migrating Images, too.